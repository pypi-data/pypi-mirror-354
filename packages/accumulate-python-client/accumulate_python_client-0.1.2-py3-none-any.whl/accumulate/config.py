# accumulate/config.py

# Define Accumulate network base URLs (without "/v3")
# ACCUMULATE_RPC_URL = "https://testnet.accumulatenetwork.io"  # Default: Testnet

ACCUMULATE_RPC_URL = "https://mainnet.accumulatenetwork.io"  # Default: Testnet

# Optional: Add mainnet and custom options
ACCUMULATE_MAINNET_URL = "https://mainnet.accumulatenetwork.io"
ACCUMULATE_TESTNET_URL = "https://testnet.accumulatenetwork.io"
ACCUMULATE_CUSTOM_URL = None  # Set this if using a custom Accumulate node

# Function to get the active Accumulate RPC URL
def get_accumulate_rpc_url():
    """Returns the currently selected Accumulate RPC URL."""
    return ACCUMULATE_CUSTOM_URL or ACCUMULATE_RPC_URL
