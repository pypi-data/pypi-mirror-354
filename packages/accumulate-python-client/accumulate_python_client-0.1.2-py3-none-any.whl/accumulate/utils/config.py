# accumulate-python-client\accumulate\utils\config.py

import os

class Config:
    """
    Configuration utility for managing environment-specific settings.
    """
    @staticmethod
    def is_testnet() -> bool:
        """
        Dynamically get the testnet status based on the environment variable.
        """
        return os.getenv("ACCUMULATE_IS_TESTNET", "false").lower() == "true"

    @staticmethod
    def initial_acme_oracle() -> float:
        """
        Dynamically get the initial ACME oracle value based on the network type.
        """
        return 5000 if Config.is_testnet() else 0.50

    @staticmethod
    def get_network_type() -> str:
        """
        Get the current network type.
        :return: 'testnet' or 'mainnet'
        """
        return "testnet" if Config.is_testnet() else "mainnet"

    @staticmethod
    def get_initial_oracle_value() -> float:
        """
        Get the initial ACME oracle value for the configured network.
        :return: The oracle value as a float.
        """
        return Config.initial_acme_oracle()
