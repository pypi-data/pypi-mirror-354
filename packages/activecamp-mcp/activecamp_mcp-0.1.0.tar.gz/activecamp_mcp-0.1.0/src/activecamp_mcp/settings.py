from pydantic_settings import BaseSettings, SettingsConfigDict


class Settings(BaseSettings):
    host: str = "0.0.0.0"
    port: int = 8080
    prometheus_port: int = 9090
    log_level: str = "INFO"
    reload: bool = False
    workers: int = 1

    mcp_dir: str = "/etc/fastmcp/"
    mcp_webhook_url: str = ""
    mcp_webhook_extra_data: dict = {}
    ac_api_url: str
    ac_api_token: str

    model_config = SettingsConfigDict(
        env_file=".env", env_file_encoding="utf-8", case_sensitive=False, extra="ignore"
    )


settings = Settings()
