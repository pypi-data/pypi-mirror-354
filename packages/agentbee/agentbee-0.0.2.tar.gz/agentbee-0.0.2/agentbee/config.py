import configparser
from pathlib import Path

CONFIG_DIR = Path.home() / ".agentbee"
CONFIG_FILE = CONFIG_DIR / "config.ini"
CONFIG_SECTION = "DEFAULT"

def get_config_path() -> Path:
    """Returns the path to the config file."""
    return CONFIG_FILE

def save_config(api_key: str, base_url: str, model: str):
    """Saves the configuration to the global config file."""
    CONFIG_DIR.mkdir(exist_ok=True)
    config = configparser.ConfigParser()
    config[CONFIG_SECTION] = {
        'llm_api_key': api_key,
        'llm_base_url': base_url,
        'llm_model': model
    }
    with open(CONFIG_FILE, 'w') as configfile:
        config.write(configfile)
    print(f"âœ… Configuration saved to {CONFIG_FILE}")

def load_config() -> dict:
    if not CONFIG_FILE.exists():
        return {}
    config = configparser.ConfigParser()
    config.read(CONFIG_FILE)
    return {
        'llm_api_key': config.get(CONFIG_SECTION, 'llm_api_key', fallback=''),
        'llm_base_url': config.get(CONFIG_SECTION, 'llm_base_url', fallback=''),
        'llm_model': config.get(CONFIG_SECTION, 'llm_model', fallback='')
    }