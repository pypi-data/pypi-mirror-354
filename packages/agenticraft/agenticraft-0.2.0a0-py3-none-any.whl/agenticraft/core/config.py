"""Configuration management for AgentiCraft.

This module provides centralized configuration management using
pydantic-settings. Configuration can be loaded from environment
variables, .env files, or set programmatically.

Example:
    Loading configuration::

        from agenticraft.core.config import settings

        # Access configuration
        print(settings.openai_api_key)
        print(settings.default_model)

        # Override configuration
        settings.default_temperature = 0.5
"""

from functools import lru_cache
from pathlib import Path
from typing import Any

from pydantic import Field, field_validator
from pydantic_settings import BaseSettings, SettingsConfigDict


class AgentiCraftSettings(BaseSettings):
    """Global settings for AgentiCraft.

    Settings are loaded from (in order of priority):
    1. Environment variables
    2. .env file in current directory
    3. .env file in parent directories
    4. Default values

    All settings can be overridden using environment variables
    with the AGENTICRAFT_ prefix.
    """

    model_config = SettingsConfigDict(
        env_prefix="AGENTICRAFT_",
        env_file=".env",
        env_file_encoding="utf-8",
        case_sensitive=False,
        extra="allow",
    )

    # General settings
    environment: str = Field(
        default="development",
        description="Environment (development, staging, production)",
    )
    debug: bool = Field(default=False, description="Enable debug mode")
    log_level: str = Field(default="INFO", description="Logging level")

    # API Keys
    openai_api_key: str | None = Field(default=None, description="OpenAI API key")
    anthropic_api_key: str | None = Field(default=None, description="Anthropic API key")
    google_api_key: str | None = Field(default=None, description="Google API key")

    # Model defaults
    default_model: str = Field(default="gpt-4", description="Default LLM model")
    default_temperature: float = Field(
        default=0.7, ge=0.0, le=2.0, description="Default temperature for LLM"
    )
    default_max_tokens: int | None = Field(
        default=None, gt=0, description="Default max tokens for LLM"
    )
    default_timeout: int = Field(
        default=30, gt=0, description="Default timeout in seconds"
    )
    default_max_retries: int = Field(default=3, ge=0, description="Default max retries")

    # Provider settings
    openai_base_url: str = Field(
        default="https://api.openai.com/v1", description="OpenAI API base URL"
    )
    anthropic_base_url: str = Field(
        default="https://api.anthropic.com", description="Anthropic API base URL"
    )
    ollama_base_url: str = Field(
        default="http://localhost:11434", description="Ollama base URL"
    )

    # Memory settings
    memory_backend: str = Field(
        default="sqlite", description="Memory backend (sqlite, json, redis)"
    )
    memory_path: Path = Field(
        default=Path("./memory"), description="Path for memory storage"
    )
    conversation_memory_size: int = Field(
        default=10, gt=0, description="Number of conversation turns to keep"
    )

    # Telemetry settings
    telemetry_enabled: bool = Field(default=True, description="Enable telemetry")
    telemetry_service_name: str = Field(
        default="agenticraft", description="Service name for telemetry"
    )
    telemetry_export_endpoint: str | None = Field(
        default=None, description="OTLP endpoint for telemetry export"
    )
    telemetry_sample_rate: float = Field(
        default=1.0, ge=0.0, le=1.0, description="Telemetry sampling rate"
    )

    # Plugin settings
    plugins_enabled: bool = Field(default=True, description="Enable plugin system")
    plugin_dirs: list[Path] = Field(
        default_factory=lambda: [Path("./plugins")],
        description="Directories to load plugins from",
    )

    # MCP settings
    mcp_enabled: bool = Field(default=True, description="Enable Model Context Protocol")
    mcp_servers: list[str] = Field(
        default_factory=list, description="MCP server URLs to connect to"
    )

    # Tool settings
    tool_execution_timeout: int = Field(
        default=10, gt=0, description="Timeout for tool execution in seconds"
    )
    tool_retry_attempts: int = Field(
        default=2, ge=0, description="Number of retry attempts for failed tools"
    )

    # Workflow settings
    workflow_step_timeout: int = Field(
        default=60, gt=0, description="Default timeout for workflow steps in seconds"
    )
    workflow_max_parallel_steps: int = Field(
        default=5, gt=0, description="Maximum parallel steps in workflows"
    )

    # Security settings
    allow_code_execution: bool = Field(
        default=False, description="Allow execution of arbitrary code in tools"
    )
    allowed_domains: set[str] = Field(
        default_factory=set, description="Allowed domains for web requests"
    )

    @field_validator("log_level")
    def validate_log_level(cls, v: str) -> str:
        """Validate log level."""
        valid_levels = {"DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"}
        v_upper = v.upper()
        if v_upper not in valid_levels:
            raise ValueError(f"Invalid log level: {v}. Must be one of {valid_levels}")
        return v_upper

    @field_validator("environment")
    def validate_environment(cls, v: str) -> str:
        """Validate environment."""
        valid_envs = {"development", "staging", "production", "test"}
        v_lower = v.lower()
        if v_lower not in valid_envs:
            raise ValueError(f"Invalid environment: {v}. Must be one of {valid_envs}")
        return v_lower

    @field_validator("memory_backend")
    def validate_memory_backend(cls, v: str) -> str:
        """Validate memory backend."""
        valid_backends = {"sqlite", "json", "redis"}
        v_lower = v.lower()
        if v_lower not in valid_backends:
            raise ValueError(
                f"Invalid memory backend: {v}. Must be one of {valid_backends}"
            )
        return v_lower

    def get_api_key(self, provider: str) -> str | None:
        """Get API key for a provider.

        Args:
            provider: Provider name (openai, anthropic, google)

        Returns:
            API key if available
        """
        key_map = {
            "openai": self.openai_api_key,
            "anthropic": self.anthropic_api_key,
            "google": self.google_api_key,
        }
        return key_map.get(provider.lower())

    def get_base_url(self, provider: str) -> str | None:
        """Get base URL for a provider.

        Args:
            provider: Provider name

        Returns:
            Base URL if available
        """
        url_map = {
            "openai": self.openai_base_url,
            "anthropic": self.anthropic_base_url,
            "ollama": self.ollama_base_url,
        }
        return url_map.get(provider.lower())

    def to_agent_config(self) -> dict[str, Any]:
        """Convert settings to agent configuration.

        Returns:
            Dictionary suitable for Agent initialization
        """
        return {
            "model": self.default_model,
            "temperature": self.default_temperature,
            "max_tokens": self.default_max_tokens,
            "timeout": self.default_timeout,
            "max_retries": self.default_max_retries,
        }

    def to_telemetry_config(self) -> dict[str, Any]:
        """Convert settings to telemetry configuration.

        Returns:
            Dictionary suitable for Telemetry initialization
        """
        return {
            "service_name": self.telemetry_service_name,
            "export_to": self.telemetry_export_endpoint,
            "enabled": self.telemetry_enabled,
            "sample_rate": self.telemetry_sample_rate,
        }

    @property
    def is_production(self) -> bool:
        """Check if running in production."""
        return self.environment == "production"

    @property
    def is_development(self) -> bool:
        """Check if running in development."""
        return self.environment == "development"

    def validate_required_keys(self, providers: list[str]) -> None:
        """Validate that required API keys are present.

        Args:
            providers: List of providers that need keys

        Raises:
            ValueError: If required keys are missing
        """
        missing = []
        for provider in providers:
            if not self.get_api_key(provider):
                missing.append(provider)

        if missing:
            raise ValueError(
                f"Missing API keys for providers: {', '.join(missing)}. "
                f"Set environment variables or update .env file."
            )


@lru_cache
def get_settings() -> AgentiCraftSettings:
    """Get the global settings instance (cached).

    Returns:
        AgentiCraftSettings instance
    """
    return AgentiCraftSettings()


# Global settings instance
settings = get_settings()


def reload_settings() -> AgentiCraftSettings:
    """Reload settings from environment.

    This clears the cache and creates a new settings instance.
    Useful for testing or when environment variables change.

    Returns:
        New AgentiCraftSettings instance
    """
    get_settings.cache_clear()
    global settings
    settings = get_settings()
    return settings


def update_settings(**kwargs: Any) -> None:
    """Update settings programmatically.

    Args:
        **kwargs: Setting values to update

    Example:
        update_settings(default_model="gpt-3.5-turbo", debug=True)
    """
    for key, value in kwargs.items():
        if hasattr(settings, key):
            setattr(settings, key, value)
        else:
            raise ValueError(f"Unknown setting: {key}")
