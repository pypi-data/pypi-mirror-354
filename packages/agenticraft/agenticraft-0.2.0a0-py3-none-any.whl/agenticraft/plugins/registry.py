"""Plugin registry for AgentiCraft.

This module provides a centralized registry for managing plugins,
their dependencies, and lifecycle. It ensures plugins are loaded
in the correct order and handles plugin interactions.

Example:
    Using the plugin registry::

        from agenticraft.plugins import PluginRegistry

        registry = PluginRegistry()

        # Register a plugin
        registry.register(weather_plugin)

        # Get tools from all plugins
        all_tools = registry.get_all_tools()

        # Get plugins by capability
        tool_plugins = registry.get_plugins_by_capability("tools")
"""

import logging
from collections import defaultdict
from typing import TYPE_CHECKING, Any

from .base import BasePlugin, PluginInfo

if TYPE_CHECKING:
    from ..core.agent import Agent
    from ..core.tool import Tool

logger = logging.getLogger(__name__)


class PluginDependencyError(Exception):
    """Raised when plugin dependencies cannot be resolved."""

    pass


class PluginRegistry:
    """Central registry for managing AgentiCraft plugins."""

    def __init__(self):
        """Initialize plugin registry."""
        # Registered plugins by name
        self._plugins: dict[str, BasePlugin] = {}

        # Plugin metadata cache
        self._plugin_info: dict[str, PluginInfo] = {}

        # Capability indexes
        self._tools_index: dict[str, list[BasePlugin]] = defaultdict(list)
        self._agents_index: dict[str, list[BasePlugin]] = defaultdict(list)
        self._providers_index: dict[str, list[BasePlugin]] = defaultdict(list)

        # Plugin dependencies
        self._dependencies: dict[str, set[str]] = defaultdict(set)
        self._dependents: dict[str, set[str]] = defaultdict(set)

        # Lifecycle tracking
        self._initialized_plugins: set[str] = set()
        self._load_order: list[str] = []

    def register(self, plugin: BasePlugin) -> None:
        """Register a plugin with the registry.

        Args:
            plugin: Plugin instance to register

        Raises:
            ValueError: If plugin with same name already registered
        """
        info = plugin.get_info()

        if info.name in self._plugins:
            raise ValueError(f"Plugin '{info.name}' already registered")

        # Store plugin and info
        self._plugins[info.name] = plugin
        self._plugin_info[info.name] = info

        # Update capability indexes
        self._update_indexes(plugin, info)

        # Track dependencies
        self._update_dependencies(info)

        # Add to load order
        self._load_order.append(info.name)

        logger.info(f"Registered plugin: {info.name} v{info.version}")

    def unregister(self, plugin: BasePlugin) -> None:
        """Unregister a plugin from the registry.

        Args:
            plugin: Plugin instance to unregister
        """
        info = plugin.get_info()

        if info.name not in self._plugins:
            return

        # Check dependents
        if info.name in self._dependents and self._dependents[info.name]:
            dependents = ", ".join(self._dependents[info.name])
            logger.warning(
                f"Unregistering plugin '{info.name}' which has dependents: {dependents}"
            )

        # Remove from indexes
        self._remove_from_indexes(plugin, info)

        # Remove from tracking
        del self._plugins[info.name]
        del self._plugin_info[info.name]

        if info.name in self._initialized_plugins:
            self._initialized_plugins.remove(info.name)

        if info.name in self._load_order:
            self._load_order.remove(info.name)

        # Clean up dependencies
        if info.name in self._dependencies:
            del self._dependencies[info.name]

        for deps in self._dependents.values():
            deps.discard(info.name)

        logger.info(f"Unregistered plugin: {info.name}")

    def _update_indexes(self, plugin: BasePlugin, info: PluginInfo) -> None:
        """Update capability indexes for a plugin."""
        # Tool index
        for tool_name in info.provides_tools:
            self._tools_index[tool_name].append(plugin)

        # Agent index
        for agent_name in info.provides_agents:
            self._agents_index[agent_name].append(plugin)

        # Provider index
        for provider_name in info.provides_providers:
            self._providers_index[provider_name].append(plugin)

    def _remove_from_indexes(self, plugin: BasePlugin, info: PluginInfo) -> None:
        """Remove plugin from capability indexes."""
        # Tool index
        for tool_name in info.provides_tools:
            if plugin in self._tools_index[tool_name]:
                self._tools_index[tool_name].remove(plugin)

        # Agent index
        for agent_name in info.provides_agents:
            if plugin in self._agents_index[agent_name]:
                self._agents_index[agent_name].remove(plugin)

        # Provider index
        for provider_name in info.provides_providers:
            if plugin in self._providers_index[provider_name]:
                self._providers_index[provider_name].remove(plugin)

    def _update_dependencies(self, info: PluginInfo) -> None:
        """Update dependency tracking for a plugin."""
        # Parse dependencies from info
        deps = set()
        if hasattr(info, "depends_on"):
            deps.update(info.depends_on)

        self._dependencies[info.name] = deps

        # Update dependents
        for dep in deps:
            self._dependents[dep].add(info.name)

    def get_plugin(self, name: str) -> BasePlugin | None:
        """Get a registered plugin by name.

        Args:
            name: Plugin name

        Returns:
            Plugin instance or None
        """
        return self._plugins.get(name)

    def get_plugin_info(self, name: str) -> PluginInfo | None:
        """Get plugin information by name.

        Args:
            name: Plugin name

        Returns:
            Plugin info or None
        """
        return self._plugin_info.get(name)

    def list_plugins(self) -> list[PluginInfo]:
        """List all registered plugins.

        Returns:
            List of plugin information
        """
        return list(self._plugin_info.values())

    def get_plugins_by_capability(self, capability: str) -> list[BasePlugin]:
        """Get plugins that provide a specific capability.

        Args:
            capability: Capability type ("tools", "agents", "providers")

        Returns:
            List of plugins with the capability
        """
        plugins = []

        if capability == "tools":
            for plugin_list in self._tools_index.values():
                for plugin in plugin_list:
                    if plugin not in plugins:
                        plugins.append(plugin)

        elif capability == "agents":
            for plugin_list in self._agents_index.values():
                for plugin in plugin_list:
                    if plugin not in plugins:
                        plugins.append(plugin)

        elif capability == "providers":
            for plugin_list in self._providers_index.values():
                for plugin in plugin_list:
                    if plugin not in plugins:
                        plugins.append(plugin)

        return plugins

    def get_all_tools(self) -> dict[str, "Tool"]:
        """Get all tools from registered plugins.

        Returns:
            Dict mapping tool names to tool instances
        """
        tools = {}

        for plugin in self._plugins.values():
            try:
                plugin_tools = plugin.get_tools()
                for tool in plugin_tools:
                    if hasattr(tool, "name"):
                        tools[tool.name] = tool
            except Exception as e:
                logger.error(f"Error getting tools from plugin '{plugin.name}': {e}")

        return tools

    def get_all_agents(self) -> dict[str, type]:
        """Get all agent classes from registered plugins.

        Returns:
            Dict mapping agent names to agent classes
        """
        agents = {}

        for plugin in self._plugins.values():
            try:
                plugin_agents = plugin.get_agents()
                for agent_class in plugin_agents:
                    agents[agent_class.__name__] = agent_class
            except Exception as e:
                logger.error(f"Error getting agents from plugin '{plugin.name}': {e}")

        return agents

    def get_all_providers(self) -> dict[str, type]:
        """Get all providers from registered plugins.

        Returns:
            Dict mapping provider names to provider classes
        """
        providers = {}

        for plugin in self._plugins.values():
            try:
                plugin_providers = plugin.get_providers()
                providers.update(plugin_providers)
            except Exception as e:
                logger.error(
                    f"Error getting providers from plugin '{plugin.name}': {e}"
                )

        return providers

    def enhance_agent(self, agent: "Agent") -> "Agent":
        """Apply all plugin enhancements to an agent.

        Args:
            agent: Agent to enhance

        Returns:
            Enhanced agent
        """
        for plugin in self._plugins.values():
            try:
                agent = plugin.enhance_agent(agent)
            except Exception as e:
                logger.error(f"Error enhancing agent with plugin '{plugin.name}': {e}")

        return agent

    def initialize_all(self) -> None:
        """Initialize all registered plugins in dependency order."""
        # Get initialization order
        init_order = self._get_initialization_order()

        for plugin_name in init_order:
            if plugin_name in self._initialized_plugins:
                continue

            plugin = self._plugins[plugin_name]

            try:
                plugin.initialize()
                self._initialized_plugins.add(plugin_name)
                logger.info(f"Initialized plugin: {plugin_name}")
            except Exception as e:
                logger.error(f"Failed to initialize plugin '{plugin_name}': {e}")
                raise

    def cleanup_all(self) -> None:
        """Clean up all plugins in reverse dependency order."""
        # Clean up in reverse order
        cleanup_order = list(reversed(self._get_initialization_order()))

        for plugin_name in cleanup_order:
            if plugin_name not in self._initialized_plugins:
                continue

            plugin = self._plugins[plugin_name]

            try:
                plugin.cleanup()
                self._initialized_plugins.remove(plugin_name)
                logger.info(f"Cleaned up plugin: {plugin_name}")
            except Exception as e:
                logger.error(f"Error cleaning up plugin '{plugin_name}': {e}")

    def _get_initialization_order(self) -> list[str]:
        """Get plugin initialization order based on dependencies.

        Returns:
            List of plugin names in initialization order

        Raises:
            PluginDependencyError: If circular dependencies detected
        """
        # Topological sort
        visited = set()
        temp_visited = set()
        order = []

        def visit(name: str):
            if name in temp_visited:
                raise PluginDependencyError(
                    f"Circular dependency detected involving '{name}'"
                )

            if name in visited:
                return

            temp_visited.add(name)

            # Visit dependencies first
            if name in self._dependencies:
                for dep in self._dependencies[name]:
                    if dep in self._plugins:  # Only if dependency is registered
                        visit(dep)

            temp_visited.remove(name)
            visited.add(name)
            order.append(name)

        # Visit all plugins
        for plugin_name in self._plugins:
            if plugin_name not in visited:
                visit(plugin_name)

        return order

    def validate_dependencies(self) -> list[str]:
        """Validate all plugin dependencies are satisfied.

        Returns:
            List of missing dependencies
        """
        missing = []

        for plugin_name, deps in self._dependencies.items():
            for dep in deps:
                if dep not in self._plugins:
                    missing.append(f"{plugin_name} requires {dep}")

        return missing

    def get_plugin_stats(self) -> dict[str, Any]:
        """Get statistics about registered plugins.

        Returns:
            Dictionary of statistics
        """
        total_tools = sum(
            len(info.provides_tools) for info in self._plugin_info.values()
        )
        total_agents = sum(
            len(info.provides_agents) for info in self._plugin_info.values()
        )
        total_providers = sum(
            len(info.provides_providers) for info in self._plugin_info.values()
        )

        return {
            "total_plugins": len(self._plugins),
            "initialized_plugins": len(self._initialized_plugins),
            "total_tools": total_tools,
            "total_agents": total_agents,
            "total_providers": total_providers,
            "unique_tools": len(self._tools_index),
            "unique_agents": len(self._agents_index),
            "unique_providers": len(self._providers_index),
            "plugins_with_dependencies": len(
                [d for d in self._dependencies.values() if d]
            ),
            "missing_dependencies": len(self.validate_dependencies()),
        }

    def clear(self) -> None:
        """Clear all registered plugins."""
        # Clean up first
        self.cleanup_all()

        # Clear all data
        self._plugins.clear()
        self._plugin_info.clear()
        self._tools_index.clear()
        self._agents_index.clear()
        self._providers_index.clear()
        self._dependencies.clear()
        self._dependents.clear()
        self._initialized_plugins.clear()
        self._load_order.clear()

        logger.info("Cleared plugin registry")


# Global registry instance
_global_registry: PluginRegistry | None = None


def get_plugin_registry() -> PluginRegistry:
    """Get the global plugin registry.

    Returns:
        Global PluginRegistry instance
    """
    global _global_registry
    if _global_registry is None:
        _global_registry = PluginRegistry()
    return _global_registry


def register_plugin(plugin: BasePlugin) -> None:
    """Register a plugin with the global registry.

    Args:
        plugin: Plugin to register
    """
    registry = get_plugin_registry()
    registry.register(plugin)


def get_all_plugin_tools() -> dict[str, "Tool"]:
    """Get all tools from registered plugins.

    Returns:
        Dict of tool name to tool instance
    """
    registry = get_plugin_registry()
    return registry.get_all_tools()
