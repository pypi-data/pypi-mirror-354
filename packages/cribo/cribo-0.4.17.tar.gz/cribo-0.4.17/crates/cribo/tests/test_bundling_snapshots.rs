#![allow(clippy::disallowed_methods)] // insta macros use unwrap internally

use anyhow::Result;
use std::fs;
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

use cribo::bundler::Bundler;
use cribo::config::Config;
use cribo::util::get_python_executable;

// Ruff linting integration for cross-validation
use ruff_linter::linter::{ParseSource, lint_only};
use ruff_linter::registry::Rule;
use ruff_linter::settings::{LinterSettings, flags};
use ruff_linter::source_kind::SourceKind;
use ruff_python_ast::PySourceType;

/// Structured execution results for better snapshot formatting
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct ExecutionResults {
    status: ExecutionStatus,
    stdout: String,
    stderr: String,
}

#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
enum ExecutionStatus {
    Success,
    Failed(i32),
}

/// Ruff linting results for cross-validation
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct RuffLintResults {
    f401_violations: Vec<String>, // Unused imports
    f404_violations: Vec<String>, // Late future imports
    other_violations: Vec<String>,
    total_violations: usize,
}

/// Run ruff linting on bundled code to cross-validate import handling
fn run_ruff_lint_on_bundle(bundled_code: &str) -> RuffLintResults {
    // Create settings for multiple import-related rules with both F401 and F404 enabled
    let settings = LinterSettings {
        rules: [Rule::UnusedImport, Rule::LateFutureImport]
            .into_iter()
            .collect(),
        ..LinterSettings::default()
    };

    let path = Path::new("<bundled>.py");
    let source_kind = SourceKind::Python(bundled_code.to_string());

    let result = lint_only(
        path,
        None,
        &settings,
        flags::Noqa::Enabled,
        &source_kind,
        PySourceType::Python,
        ParseSource::None,
    );

    let mut f401_violations = Vec::new();
    let mut f404_violations = Vec::new();
    let mut other_violations = Vec::new();

    for message in &result.messages {
        if let Some(rule) = message.to_rule() {
            let location = message.compute_start_location();
            let violation_info = format!(
                "Line {}: {} - {}",
                location.line.get(),
                rule.noqa_code(),
                message.body()
            );

            match rule {
                Rule::UnusedImport => f401_violations.push(violation_info),
                Rule::LateFutureImport => f404_violations.push(violation_info),
                _ => other_violations.push(violation_info),
            }
        }
    }

    let total_violations = f401_violations.len() + f404_violations.len() + other_violations.len();

    RuffLintResults {
        f401_violations,
        f404_violations,
        other_violations,
        total_violations,
    }
}

/// Generic test that processes all fixture directories in tests/fixtures/bundling
/// Each directory should contain a main.py entry point and will be bundled and executed
#[test]
fn test_all_bundling_fixtures() -> Result<()> {
    let fixtures_dir = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("fixtures")
        .join("bundling");

    // Skip if bundling directory doesn't exist
    if !fixtures_dir.exists() {
        return Ok(());
    }

    // Read all subdirectories in the bundling fixtures directory
    let entries = fs::read_dir(&fixtures_dir)?;
    let mut fixture_names = Vec::new();

    for entry in entries {
        let entry = entry?;
        let path = entry.path();

        if path.is_dir() {
            if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                fixture_names.push(name.to_string());
            }
        }
    }

    // Sort for deterministic test order
    fixture_names.sort();

    // Process each fixture directory
    for fixture_name in fixture_names {
        test_single_bundling_fixture(&fixtures_dir, &fixture_name)?;
    }

    Ok(())
}

/// Test a single bundling fixture directory
fn test_single_bundling_fixture(fixtures_dir: &Path, fixture_name: &str) -> Result<()> {
    let fixture_path = fixtures_dir.join(fixture_name);
    let main_py_path = fixture_path.join("main.py");

    // Skip if main.py doesn't exist
    if !main_py_path.exists() {
        // Skip silently or use conditional verbose logging
        if std::env::var("RUST_TEST_VERBOSE").is_ok() {
            eprintln!("Skipping {}: no main.py found", fixture_name);
        }
        return Ok(());
    }

    // Create temporary directory for output
    let temp_dir = TempDir::new()?;
    let bundle_path = temp_dir.path().join("bundled.py");

    // Configure bundler
    let config = Config::default();
    let mut bundler = Bundler::new(config);

    // Bundle the fixture
    bundler.bundle(&main_py_path, &bundle_path, false)?;

    // Optionally validate Python syntax before execution
    let python_cmd = get_python_executable();
    let syntax_check = Command::new(&python_cmd)
        .args(["-m", "py_compile"])
        .arg(&bundle_path)
        .output();
    if let Ok(output) = syntax_check {
        if !output.status.success() && std::env::var("RUST_TEST_VERBOSE").is_ok() {
            eprintln!(
                "Warning: Bundled code has syntax errors for fixture {}",
                fixture_name
            );
            eprintln!("Stderr: {}", String::from_utf8_lossy(&output.stderr));
        }
    }

    // Read the bundled code and normalize line endings for cross-platform compatibility
    let bundled_code = fs::read_to_string(&bundle_path)?;

    // Run ruff linting for cross-validation of unused imports elimination
    let ruff_results = run_ruff_lint_on_bundle(&bundled_code);

    // Execute the bundled code with Python and capture output
    // Use Python executable with virtual environment support
    let python_cmd = get_python_executable();

    let python_output = Command::new(&python_cmd)
        .arg(&bundle_path)
        .current_dir(temp_dir.path()) // Set working directory for consistent execution
        .output()
        .map_err(|e| {
            anyhow::anyhow!(
                "Failed to execute Python: {} (command: {} {:?})",
                e,
                python_cmd,
                bundle_path
            )
        })?;

    // Note: For timeout support, we'd need a more complex solution or external crate
    // The standard library doesn't provide built-in timeout for process execution

    // Create separate snapshots using insta's named snapshot feature
    insta::with_settings!({
        snapshot_suffix => fixture_name,
        omit_expression => true
    }, {
        // Snapshot the bundled code
        insta::assert_snapshot!("bundled_code", bundled_code);

        // Create structured execution results snapshot
        let execution_status = if python_output.status.success() {
            ExecutionStatus::Success
        } else {
            ExecutionStatus::Failed(python_output.status.code().unwrap_or(-1))
        };

        let execution_results = ExecutionResults {
            status: execution_status,
            stdout: String::from_utf8_lossy(&python_output.stdout)
                .trim()
                .replace("\r\n", "\n")
                .to_string(),
            stderr: String::from_utf8_lossy(&python_output.stderr)
                .trim()
                .replace("\r\n", "\n")
                .to_string(),
        };

        insta::assert_debug_snapshot!("execution_results", execution_results);

        // Snapshot ruff linting results for cross-validation
        insta::assert_debug_snapshot!("ruff_lint_results", ruff_results);
    });

    Ok(())
}
