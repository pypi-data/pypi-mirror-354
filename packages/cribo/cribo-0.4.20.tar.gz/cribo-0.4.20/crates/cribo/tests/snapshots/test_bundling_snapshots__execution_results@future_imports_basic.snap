---
source: crates/cribo/tests/test_bundling_snapshots.rs
---
ExecutionResults {
    status: Success,
    stdout: "Processing result: {'input': {'key': 'value', 'numbers': [1, 2, 3]}, 'processed': True, 'output': ['key=value', 'numbers=[1, 2, 3]']}",
    stderr: "",
}
